### DISCUSSION TITLE
LPS swarm with one drone not receiving correct position

### DISCUSSION THREAD

**Discussion by ErnstKapp** [bitcraze/discussions#1418] [Q&A] [answered]
Hello!

I've been trying to implement a GNSS based collision avoidance program on two drones using the swarm library and the Loco positioning system. One of the drones is receiving positional data as it should and has no problem flying the intended route. However, the second drone in the swarm is constantly getting nonsense positional data and can not fly properly. I do not believe it to be a hardware issue since switching addresses and such on the drones yield the same malfunction (also the anchors are in TDoA 2 mode). Is it the way that i retrieve positional data in swarm._cfs.items() is not threaded maybe? Some of the code is currently hardcoded to ensure that no complete nonsense positional data comes in (like x = 100  or the likes) and the crash logic is pretty jerry-rigged, but commented out atm. Any help is much appreciated :) 
          
          URI1 = 'radio://0/80/2M/E7E7E7E7E8'
          URI2 = 'radio://0/80/2M/E7E7E7E7E7'
          
          
          def log_position(scf):
              lg_stab = LogConfig(name='position', period_in_ms=30)
              lg_stab.add_variable('stateEstimate.x', 'float')
              lg_stab.add_variable('stateEstimate.y', 'float')
              lg_stab.add_variable('stateEstimate.z', 'float')
          
              uri = scf.cf.link_uri
              with SyncLogger(scf, lg_stab) as logger:
                  for log_entry in logger:
                      data = log_entry[1]
          
                      x = data['stateEstimate.x']
                      y = data['stateEstimate.y']
                      z = data['stateEstimate.z']
          
                      
                      
                      print(uri, "is at", x, y, z)
                      break
          
          def initial_position_estimator(scf):
              print('Waiting for estimator to find position...')
          
              logconf = LogConfig(name='Position', period_in_ms=10)
              logconf.add_variable('stateEstimate.x', 'float')
              logconf.add_variable('stateEstimate.y', 'float')
              logconf.add_variable('stateEstimate.z', 'float')
          
              var_y_history = [1000] * 10
              var_x_history = [1000] * 10
              var_z_history = [1000] * 10
          
              threshold = 0.5
          
              with SyncLogger(scf, logconf) as logger:
                  for log_entry in logger:
                      data = log_entry[1]
          
                      var_x_history.append(data['stateEstimate.x'])
                      var_x_history.pop(0)
                      var_y_history.append(data['stateEstimate.y'])
                      var_y_history.pop(0)
                      var_z_history.append(data['stateEstimate.z'])
                      var_z_history.pop(0)
          
                      min_x = min(var_x_history)
                      max_x = max(var_x_history)
                      min_y = min(var_y_history)
                      max_y = max(var_y_history)
                      min_z = min(var_z_history)
                      max_z = max(var_z_history)
          
                      if (max_x - min_x) < threshold and (
                              max_y - min_y) < threshold and (
                              max_z - min_z) < threshold and (
                                  abs(min_x) < 10 ) and ( 
                                  abs( min_y ) < 10 ) and (
                                  abs(min_z) < 0.8):
                          print("{} {} {}".
                             format(min_x, min_y, min_z))
                          return min_x, min_y
                          
          
          def wait_for_position_estimator(scf):
              print('Waiting for estimator...')
          
              logconf = LogConfig(name='Position', period_in_ms=10)
              logconf.add_variable('stateEstimate.x', 'float')
              logconf.add_variable('stateEstimate.y', 'float')
              logconf.add_variable('stateEstimate.z', 'float')
          
              var_y_history = [1000] * 2
              var_x_history = [1000] * 2
              var_z_history = [1000] * 2
          
              threshold = 0.2
          
              with SyncLogger(scf, logconf) as logger:
                  
                  for log_entry in logger:
          
                      data = log_entry[1]
                      
                      var_x_history.append(data['stateEstimate.x'])
                      var_x_history.pop(0)
                      var_y_history.append(data['stateEstimate.y'])
                      var_y_history.pop(0)
                      var_z_history.append(data['stateEstimate.z'])
                      var_z_history.pop(0)
          
                      min_x = min(var_x_history)
                      max_x = max(var_x_history)
                      min_y = min(var_y_history)
                      max_y = max(var_y_history)
                      min_z = min(var_z_history)
                      max_z = max(var_z_history)
          
                      # print("{} {} {}".
                      #       format(max_x - min_x, max_y - min_y, max_z - min_z))
          
                      if (max_x - min_x) < threshold and (
                              max_y - min_y) < threshold and (
                              max_z - min_z) < threshold and (
                                  abs(min_x) < 10 ) and ( 
                                  abs( min_y ) < 10 ) and (
                                  abs( max_z ) < 5):
                          print("{} {} {}".
                             format(min_x, min_y, min_z))
                          return min_x, min_y, min_z
                     
          
          def activate_high_level_commander(cf):
              cf.param.set_value('commander.enHighLevel', '1')
          
          def up(scf):
              cf = scf.cf
              cf.commander.send_velocity_world_setpoint(0, 0, 2.0, 0.0)
              time.sleep(1.0)
              
          def stop(scf, r, alpha):
              cf = scf.cf
              cf.commander.send_velocity_world_setpoint(0, 0, -0.5, 0.0)
                  
          def find_target(scf, target, drone):
              
                  cf = scf.cf
                  
                  drone_x, drone_y = drone[0], drone[1]
                  target_x, target_y = target[0], target[1]
          
                  relative_pos_x = target_x - drone_x
                  relative_pos_y = target_y - drone_y
                  normalized_velocity_x = 0.3 * relative_pos_x / (abs(relative_pos_x) + abs(relative_pos_y))
                  normalized_velocity_y = 0.3 * relative_pos_y / (abs(relative_pos_x) + abs(relative_pos_y))
          
                  if drone[2] < 1:
                      cf.commander.send_velocity_world_setpoint(normalized_velocity_x, normalized_velocity_y, 1.0, 0.0)
                  elif drone[2] > 2:
                      cf.commander.send_velocity_world_setpoint(normalized_velocity_x, normalized_velocity_y, -0.3, 0.0)
                  else:
                      cf.commander.send_velocity_world_setpoint(normalized_velocity_x, normalized_velocity_y, 0.0, 0.0)
          
          
          def reset_estimator(scf):
              scf.cf.param.set_value('kalman.resetEstimation', '1')
              time.sleep(0.1)
              scf.cf.param.set_value('kalman.resetEstimation', '0')
          
              wait_for_position_estimator(scf)
          
          def log_pos_callback(timestamp, data, logconf):
              return data
          
          def add_log_config(scf, logconf):
              scf.cf.log.add_config(logconf)
          
          def distance_between(drone1, drone2):
                  delta_x = abs(drone1[0] - drone2[0])
                  delta_y = abs(drone1[1] - drone2[1])
                  print("distance between drones:", math.sqrt(delta_x ** 2 + delta_y ** 2))
                  return math.sqrt(delta_x ** 2 + delta_y ** 2)
          
          def avoidance_control(scf, drone_velocity, drone_pos, obstacle_pos):
             
              cf = scf.cf
              relative_pos_x = obstacle_pos[0] - drone_pos[0]
              relative_pos_y = obstacle_pos[1] - drone_pos[1]
          
              force_x_norm =  -0.1 * relative_pos_x / (abs(relative_pos_x) + abs(relative_pos_y))
              force_y_norm =  -0.1 * relative_pos_y / (abs(relative_pos_x) + abs(relative_pos_y))
              print("repulsion")
              cf.commander.send_velocity_world_setpoint(force_x_norm, force_y_norm, 0.0, 0.0)
          
          
          def calculate_normalized_velocity(target, drone):
              relative_position_x = target[0] - drone[0]
              relative_position_y = target[1] - drone[1]
              normalized_velocity_x = 0.3 * relative_position_x / (abs(relative_position_x) + abs(relative_position_y))
              normalized_velocity_y = 0.3 * relative_position_y / (abs(relative_position_x) + abs(relative_position_y))
              return normalized_velocity_x, normalized_velocity_y
          
          
          if __name__ == '__main__':
              cflib.crtp.init_drivers()
              uris = [URI1, URI2]
              
              crashed_drone_set = set([])
          
              factory = CachedCfFactory(rw_cache='./cache')
              with Swarm(uris, factory=factory) as swarm:
          
                  
          
                  pos_params = {
                  URI1: [1.0, 2.0],
                  URI2: [-1.0, 0.0]
                  }
          
                  threshold = 0.3
                  
                  drone_pos_list = []
                  drone_target_list = []
          
                  for uri, scf in swarm._cfs.items():
                      #print(scf)
                      drone_pos_list.append(initial_position_estimator(scf))
                      #print(drone_pos_list)
          
                  for key, value in pos_params.items():
                      drone_target_list.append(value)
                  
          
                  swarm.parallel_safe(up)
                  
                  while True:
          
                      all_within_threshold = True
                      for drone_pos, drone_target in zip(drone_pos_list, drone_target_list):
                          if abs(drone_pos[0] - drone_target[0]) > threshold or abs(drone_pos[1] - drone_target[1]) > threshold:
                              all_within_threshold = False
                              break
                      
                      drone_list = []
                      
                      for uri, scf in swarm._cfs.items():
                          IsCrashed = False
                          position = wait_for_position_estimator(scf)
                          if math.sqrt(position[0]**2 + position[1]**2) < 0.02 or uri in crashed_drone_set:
                              IsCrashed = True
                              print("CRASH")
                              if uri not in crashed_drone_set:
                                  crashed_drone_set.add(uri)
          
                          drone_list.append([uri, position, IsCrashed])
                      
                      
                      if distance_between(drone_list[0][1], drone_list[1][1]) < 0.6 and not drone_list[0][2] and \
                          not drone_list[1][2]:
                          repulsion_params = {
                              URI1: [calculate_normalized_velocity(drone_target_list[0], drone_list[0][1]),
                                    drone_list[0][1], drone_list[1][1]], 
                              URI2: [calculate_normalized_velocity(drone_target_list[1], drone_list[1][1]),
                                    drone_list[1][1], drone_list[0][1]]}
                          swarm.parallel_safe(avoidance_control, args_dict = repulsion_params)
                      
                      else:
                          position_params = {
                              URI1: [drone_target_list[0],
                                  drone_list[0][1] ],
                              URI2: [drone_target_list[1],
                                  drone_list[1][1] ]
                              }
                          swarm.parallel_safe(find_target, args_dict = position_params)
                          
                      if all_within_threshold:
                          swarm.parallel(stop)
                          break

**Comment by knmcguire** [comment]
Hi! Do you know which loco positioning method you are using? TWR is technically only meant for one crazyflie, not more. So you should probably switch to TDOA2 or 3

See this part of the tutorial: https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/#switching-system-mode-to-tdoa

**Reply by ErnstKapp** [reply, marked as answer]
Yes it is in TDoA 2, just solved it an hour ago though, i changed from sending cf.commander.send_velocity_world_setpoint(0, 0, 2.0, 0.0) in the beginning for the take off to cf.commander.send_setpoint(*thrust*) and it works. Not sure why it messed with the positional coordinates but it may be some other hardware related thing with the anchors that i overlooked for some reason in prior tests. Thanks for the reply anyways!

**Reply by knmcguire** [reply]
Hmmm.... that is very strange. And how about you use the high level commander with a take off command? https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/high_level_commander/

It would make more sense that there will be some issue with the setpoint structure than the positioning as those should be completely separate frameworks within the firmware.
